### PR TITLE
[Main Course] - Feature engineering

### PR DESCRIPTION
**Pull request by jz263**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] This change requires a documentation update

# How Has This Been Tested?

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: docs/Main_Course/Data_Analytics/2021-09-25-Feature-engineering.md [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+## 1 前言
+
+前一阵子总结了下自己参加的信贷违约风险预测比赛的数据处理和建模的流程，发现自己对业务上的特征工程认识尚浅，凑巧在Kaggle上曾经也有一个金融风控领域——[房贷违约风控](https://www.kaggle.com/c/home-credit-default-risk/overview)的比赛，里面有许多大神分享了他们的特征工程方法，细看下来有不少值得参考和借鉴的地方。
+
+## 2 赛题和数据简介
+
+这个比赛也是经典的监督学习中的二分类问题，需要我们根据用户的申请信息，征信信息（Bureau）以及用户在该机构的信用历史等信息，预测申请人贷款违约的概率。由于赛题是做贷前预测，所以需要找的特征主要是挖掘客户是否存在欺诈，对于非欺诈用户，他们是否有能力还款、
+
+![](https://files.mdnice.com/user/15233/3a541655-3d6a-457f-a195-5ef56d8046d5.png)
```
Comment (jhqian0909) [inline docs/Main_Course/Data_Analytics/2021-09-25-Feature-engineering.md:18]: 最好采用之前图片的html的写法

File: docs/Main_Course/Data_Analytics/2021-09-25-Feature-engineering.md [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+自动化的数据工程这一步，有些选手几乎完全依赖自动特征工程，例如构造polynomial features，有些利用开源的数据工程包如Featuretools。
+
+## 2.1.1 近期特征
+
+有时数据中的时间信息为时间戳，我们可以手工将其转换成数值信息，如计算最近一次使用信用卡的时间，上一次联系时间，上次逾期时间等。
+
+## 2.1.2 统计特征 
+
+一个申请id会对应拉取其申请人的征信(Bureau Credit)数据和信用卡使用数据，一个申请id在其他表种对对应多行记录。对于这部分变量，往往会考虑构造统计性特征，如均值,最大/最小值,合计值，频次等等。
```
Comment (jhqian0909) [inline docs/Main_Course/Data_Analytics/2021-09-25-Feature-engineering.md:40]: 中英文的逗号

Comment (jhqian0909) [reply docs/Main_Course/Data_Analytics/2021-09-25-Feature-engineering.md:40]: 其他表中

File: docs/Main_Course/Data_Analytics/2021-09-25-Feature-engineering.md [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+* Wrapper(包装法)：根据目标函数（往往是预测效果评分），每次选择若干特征，或者排除若干特征，常见如递归特征消除法。
+* Embedded(嵌入法)：先使用某些机器学习的模型(常见用树模型)进行训练，得到各个特征的权值系数，根据系数从大到小选择特征（类似于Filter，只不过系数是通过训练得来的。
+
+为了保证模型的可解释性，例如PCA对特征做变换的特征降维方法一般不用于风控建模中。多数特征筛选方法是先用filter法移除共线特征，然后利用嵌入法计算特征重要性对进行排序。
+
+在 [Introduction to Feature Selection](https://www.kaggle.com/willkoehrsen/introduction-to-feature-selection) 中，作者最终从1465个变量中筛选342个变量放入lgb模型中，测试集AUC仅从0.783降低到0.782，未经过特征工程的lgb模型baseline为0.735。
+
+## 3.1 小结
+
+Kaggle 大神们的特征工程思路后，最大的感受是工程量很大。一些公司会选择直接将特征扩展的步骤合并在数据仓库的ETL中，从而降低建模的周期和成本。所以对于个人来说，了解数据仓库中的ETL过程，或许提高业务知识认知的手段之一。
```
Comment (jhqian0909) [inline docs/Main_Course/Data_Analytics/2021-09-25-Feature-engineering.md:101]: 这两句话似乎有点语法问题

File: docs/Main_Course/Data_Analytics/2021-09-25-Feature-engineering.md [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+例如如果一个用户在本月消费金额远高于过去12个月内的月均消费金额，可能说明客户在恶意透支信用额度或者盗刷，用户的风险应提高。
+
+## 2.1.4 特征筛选
+
+根据特征选择的形式，可分为三大类：
+
+* Filter(过滤法)：按照发散性或相关性对各个特征进行评分，设定阈值或者待选择特征的个数进行筛选，常见的指标有pearson相关系数，卡方验证，互信息等。
+* Wrapper(包装法)：根据目标函数（往往是预测效果评分），每次选择若干特征，或者排除若干特征，常见如递归特征消除法。
+* Embedded(嵌入法)：先使用某些机器学习的模型(常见用树模型)进行训练，得到各个特征的权值系数，根据系数从大到小选择特征（类似于Filter，只不过系数是通过训练得来的。
```
Comment (jhqian0909) [inline docs/Main_Course/Data_Analytics/2021-09-25-Feature-engineering.md:93]: 缺了一个括号

File: docs/Main_Course/Data_Analytics/2021-09-25-Feature-engineering.md (hunk 2 above)
Comment (jz263) [reply docs/Main_Course/Data_Analytics/2021-09-25-Feature-engineering.md]: 已修改

File: docs/Main_Course/Data_Analytics/2021-09-25-Feature-engineering.md (hunk 1 above)
Comment (jz263) [reply docs/Main_Course/Data_Analytics/2021-09-25-Feature-engineering.md]: 已修改

File: docs/Main_Course/Data_Analytics/2021-09-25-Feature-engineering.md (hunk 3 above)
Comment (jz263) [reply docs/Main_Course/Data_Analytics/2021-09-25-Feature-engineering.md]: 是有点拗口。。稍微修改了一下

File: docs/Main_Course/Data_Analytics/2021-09-25-Feature-engineering.md (hunk 4 above)
Comment (jz263) [reply docs/Main_Course/Data_Analytics/2021-09-25-Feature-engineering.md]: 已修改

**Review by jhqian0909 — APPROVED** [review]

File: docs/Main_Course/Data_Analytics/2021-09-25-Feature-engineering.md [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+
+* 申请表:贷款申请信息，主表，一行代表一个贷款申请id。
+* 征信(Bureau)余额表: Bureau信用记录，每行是一个申请用户的月度数据，一个id最多有近96个月的记录
+* 申请历史表: 同一用户的历史贷款申请信息，额度，期限，利率，是否审批通过等
+* 还款记录表: 同一用户的历史还款行为记录。
+* 信用卡余额记录表: 持有信用卡的用户的信用卡消费行为记录数据。
+
+## 2.1 特征工程
+
+特征工程的主要思路是尽可能多的构造大量特征，再利用特征筛选指标或是模型减少特征数量。
```
Comment (Realvincentyuan) [inline docs/Main_Course/Data_Analytics/2021-09-25-Feature-engineering.md:34]: 尽可能多·地·

File: docs/Main_Course/Data_Analytics/2021-09-25-Feature-engineering.md [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+
+自动化的数据工程这一步，有些选手几乎完全依赖自动特征工程，例如构造polynomial features，有些利用开源的数据工程包如Featuretools。
+
+## 2.1.1 近期特征
```
Comment (Realvincentyuan) [inline docs/Main_Course/Data_Analytics/2021-09-25-Feature-engineering.md:38]: 标题应该用###

File: docs/Main_Course/Data_Analytics/2021-09-25-Feature-engineering.md [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+
+根据特征选择的形式，可分为三大类：
+
+* Filter(过滤法)：按照发散性或相关性对各个特征进行评分，设定阈值或者待选择特征的个数进行筛选，常见的指标有pearson相关系数，卡方验证，互信息等。
+* Wrapper(包装法)：根据目标函数（往往是预测效果评分），每次选择若干特征，或者排除若干特征，常见如递归特征消除法。
+* Embedded(嵌入法)：先使用某些机器学习的模型(常见用树模型)进行训练，得到各个特征的权值系数，根据系数从大到小选择特征（类似于Filter，只不过系数是通过训练得来的）。
+
+为了保证模型的可解释性，例如PCA对特征做变换的特征降维方法一般不用于风控建模中。多数特征筛选方法是先用filter法移除共线特征，然后利用嵌入法计算特征重要性对进行排序。
+
+在 [Introduction to Feature Selection](https://www.kaggle.com/willkoehrsen/introduction-to-feature-selection) 中，作者最终从1465个变量中筛选342个变量放入lgb模型中，测试集AUC仅从0.783降低到0.782，未经过特征工程的lgb模型baseline为0.735。
```
Comment (Realvincentyuan) [inline docs/Main_Course/Data_Analytics/2021-09-25-Feature-engineering.md:101]: 感觉有的缩写第一次用的时候可以先把全称写一下

**Review by Realvincentyuan — APPROVED** [review]

File: docs/Main_Course/Data_Analytics/2021-09-25-Feature-engineering.md (hunk 6 above)
Comment (jz263) [reply docs/Main_Course/Data_Analytics/2021-09-25-Feature-engineering.md]: 已修改

File: docs/Main_Course/Data_Analytics/2021-09-25-Feature-engineering.md (hunk 7 above)
Comment (jz263) [reply docs/Main_Course/Data_Analytics/2021-09-25-Feature-engineering.md]: 已修改

File: docs/Main_Course/Data_Analytics/2021-09-25-Feature-engineering.md (hunk 5 above)
Comment (jz263) [reply docs/Main_Course/Data_Analytics/2021-09-25-Feature-engineering.md]: 已修改

**Pull request merged**